### PR TITLE
Assignment 3

### PR DESCRIPTION
**Pull request by JerettLatimer**

### CODE REVIEW

**Comment by StevenZuelke** [comment]
I'm reviewing this

File: README.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
 # Assignment
 
+[![Build Status](https://dev.azure.com/jlatimer1/CSCD379-2020-Winter/_apis/build/status/JerettLatimer.EWU-CSCD379-2020-Winter?branchName=Assignment3)](https://dev.azure.com/jlatimer1/CSCD379-2020-Winter/_build/latest?definitionId=1&branchName=Assignment3)
+
```
Comment (StevenZuelke) [inline README.md]: Badge Works

File: SecretSanta/src/SecretSanta.Data/FingerPrintEntityBase.cs [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
-        public string ModifiedBy { get; set; } = "";
-        public DateTime ModifiedOn { get; set; }
+        [Required]
+        public string? CreatedBy { get; internal set; }
+        [Required]
+        public DateTime CreatedOn { get; internal set; }
+        [Required]
+        public string? ModifiedBy { get;  internal set; }
+        [Required]
+        public DateTime ModifiedOn { get; internal set; }
```
Comment (StevenZuelke) [inline SecretSanta/src/SecretSanta.Data/FingerPrintEntityBase.cs]: Appropriately encapsulated

File: SecretSanta/src/SecretSanta.Data/EntityBase.cs [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
 using System.Text;
 
 namespace SecretSanta.Data
 {
     public class EntityBase
     {
+        [Required]
         public int Id { get; set; }
```
Comment (StevenZuelke) [inline SecretSanta/src/SecretSanta.Data/EntityBase.cs]: Believe that this should be a nullable field and protected set, this is where encapsulation fails

File: SecretSanta/src/SecretSanta.Data/Gift.cs [hunk 4]
```diff
@@ -16,5 +16,22 @@ public class Gift : FingerPrintEntityBase
...
+            User = user;
+        }
+
+        private Gift(string title, string url, string description, int userId)
+        {
+            Title = title;
+            Url = url;
+            Description = description;
+            UserId = userId;
+        }
```
Comment (StevenZuelke) [inline SecretSanta/src/SecretSanta.Data/Gift.cs]: Non default constructors added. Well done

File: SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs [hunk 5]
```diff
@@ -31,39 +25,78 @@ public async Task Gift_CanBeSavedToDatabase()
...
-            {
-                Description = null!
-            };
+            _ = new Gift(SampleData.ArduinoTitle, SampleData.ArduinoUrl, null!, SampleData.CreateJackRyan());
         }
 
         [TestMethod]
         [ExpectedException(typeof(ArgumentNullException))]
         public void Gift_SetUrlToNull_ThrowsArgumentNullException()
         {
```
Comment (StevenZuelke) [inline SecretSanta/test/SecretSanta.Data.Tests/GiftTests.cs]: You did not add the missing CRUD tests. Missing Update and Delete.

File: SecretSanta/src/SecretSanta.Business/AutomapperConfigurationProfile.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+        static public IMapper CreateMapper()
+        {
+            var mapperConfig = new MapperConfiguration(cfg =>
+            {
+                cfg.AddMaps(Assembly.GetExecutingAssembly());
+            });
+            return mapperConfig.CreateMapper();
+        }
+    }
```
Comment (StevenZuelke) [inline SecretSanta/src/SecretSanta.Business/AutomapperConfigurationProfile.cs]: Automapper done correctly

File: SecretSanta/src/SecretSanta.Business/GiftService.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    public class GiftService : EntityService<Gift>, IEntityService<Gift>
+    {
+        public GiftService(ApplicationDbContext dbContext, IMapper mapper) : base(dbContext, mapper)
+        {
+
+        }
+
+        public override async Task<Gift> FetchByIdAsync(int id) =>
+            await ApplicationDbContext.Set<Gift>().Include(nameof(Gift.User)).SingleAsync(item => item.Id == id);
+    }
```
Comment (StevenZuelke) [inline SecretSanta/src/SecretSanta.Business/GiftService.cs]: Includes User field. You are missing IGiftService though, as well as IGroupService, and you could have included the other CRUD operations like fetchall, update, delete

**Comment by MarkMichaelis** [comment]
# Assignment 3

## Main Assignment

- Connect Azure DevOps up for your fork.
- Add a Status Badge (see [Add a status badge to your repository](https://docs.microsoft.com/en-us/azure/devops/pipelines/create-first-pipeline?view=azure-devops&tabs=browser%2Ctfs-2018-2#add-a-status-badge-to-your-repository)) to the `README.md` file in `master` that shows the build status on each branch. ✔
- Be sure that clicking on the badge successfully navigates one to the Azure DevOps pipeline. ✔
- Rebase to the Readme.md update on `master` to the `Assignment3` branch so it too shows the badge (be sure it points to the Assignment3 branch status) ✔
- Appropriately encapsulate Id and all fingerprint data. ✔
- Add non-default constructors to Gift and address all warnings (see BlogEngine's Post.cs file). ✔
- Add any missing entity framework CRUD unit tests for `Gift`. ✔
- Implement an Automapper Configuration between two `Gift` objects. ✔
  - Do not map the Id ✔
- Create services for (do not map primary key Id for all entities):
  - Groups ✔
  - Users ✔
  - Gifts ✔
  - When retrieving the Gift object from the service, User should be pre-populated (included) ✔
- Refactor out SampleData (see BlogEngine's SampleData.cs) ✔

## Extra Credit

The following are options for extra credit (you don't need to do them all):

- Write exceptional code that is clean, conforms to all coding guidelines, is well refactored, and easy to understand.

## Project

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln |  |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln |  |  | 4 |


| Project Analysis |   |
| -----------------| - |
| Build/Compile | (Build succeeded: 0 Error(s), 0 Warning(s)) |
| Disabled Warnings **** | **✔**
| Nullable Enabled: | **✔** |
| Appropriately handle all warnings | **✔** |
| Test Run Succeeded: Total tests=23, Passed=23, Failed=0, Skipped=0 |

**Pull request closed without merging**